### PR TITLE
Add Lavamoat to build system

### PR DESCRIPTION
**Pull request by EtDu**
This PR aims to add LavaMoat in the MetaMask build system.

- Introduces patch-package to work around SES compatibility issues in deep dependencies
- Add explicit imports for babel plugins and presets
- Browserify: use explicit require statements to import plugins & transforms
- Browserify: use explicit imports for transforms in dependencies (required for static analysis)
- Remove redundant gulp-sass assignment to compiler

Regular config modifications:
- Fix deep endowments bug: remove .match from process.version
- Remove .call from hasOwnProperty & stream.Stream
    Lavamoat fails to dedupe configuration when merging override configuration (E.G. x.y & x.y.z)

Override config additions:
- Endow babel presets & plugins manually (dynamic require)
- Endow to @babel/core for reading babel config
- Endow Browerify transform loose-envify
- Endow post-css iterations to stylelint (lazy-import)
- incomplete parsing: vinyl, colorette, node-sass

### CODE REVIEW

**Comment by kumavis** [comment]
wow nice, this is ready

**Comment by kumavis** [comment]
this pr is ready but i recommend waiting til after the holiday

**Comment by Gudahtt** [comment]
Now that v9 is out, I think we're ready to start trying to land this. Could you rebase this on `develop`?

**Comment by kumavis** [comment]
🦇 🦇 🦇 🦇 🦇 🦇 🦇 🦇 🦇 🦇 
```
build completed. task timeline:
█ clean NaNs
█ styles:prod NaNs
█ manifest:prod NaNs
█ scripts:core:prod:disable-console NaNs
█ scripts:core:prod:initSentry NaNs
█ scripts:core:prod:phishing-detect NaNs
█ scripts:core:prod:contentscript NaNs
█ scripts:deps:ui NaNs
█ scripts:deps:background NaNs
█ static:prod NaNs
█ scripts:core:prod:background NaNs
█ scripts:core:prod:ui NaNs
█ zip NaNs
█ prod NaNs
```

**Comment by kumavis** [comment]
- [ ] de-conflict
- [ ] fix task timers
- [ ] fix lavamoat-viz
- [ ] attempt to minimalize PR diffs

**Comment by kumavis** [comment]
just did a manual squash + rebase

**Comment by kumavis** [comment]
timers fixed :+1: 
```
build completed. task timeline:
█ clean 0.0s
█████████████████████████████████████████ prod 136.2s
████▋ styles:prod 12.2s
   ▐███████████████████████████▊ scripts:deps:background 92.5s
   ▐████████████▋ scripts:deps:ui 40.9s
   ▐█████████████████████████████ scripts:core:prod:background 96.5s
   ▐████████████████████████████████████▎ scripts:core:prod:ui 121.2s
   ▐████████████▌ scripts:core:prod:phishing-detect 40.5s
   ▐████▉ scripts:core:prod:initSentry 14.5s
   ▐████████████▌ scripts:core:prod:contentscript 40.5s
   ▐███▌ scripts:core:prod:disable-console 9.5s
   ▐███████████████████████████▉ static:prod 92.5s
   ▐▋ manifest:prod 0.0s
                                       ██ zip 2.7s
Done in 144.45s.
```

**Comment by kumavis** [comment]
@EtDu looks like wrong graph, looks like the example data https://243469-42009758-gh.circle-artifacts.com/0/build-artifacts/build-viz/index.html

File: patches/sass+1.32.4.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
+diff --git a/node_modules/sass/sass.dart.js b/node_modules/sass/sass.dart.js
+index fedd867..fef6a8f 100644
+--- a/node_modules/sass/sass.dart.js
++++ b/node_modules/sass/sass.dart.js
+@@ -16,6 +16,9 @@ self.scheduleImmediate = self.setImmediate
+ // CommonJS globals.
+ self.exports = exports;
+ 
++exports.bridgeJson = (target) => JSON.parse(JSON.stringify(target))
```
Comment (Gudahtt) [inline patches/sass+1.32.4.patch:9]: Huh. That's an interesting solution.

File: package.json [hunk 2]
```diff
@@ -326,7 +332,8 @@
       "github:assemblyscript/assemblyscript": false,
       "tiny-secp256k1": false,
       "@lavamoat/preinstall-always-fail": false,
-      "fsevents": false
+      "fsevents": false,
+      "metamask-crx": false
```
Comment (Gudahtt) [inline package.json:336]: This was necessitated by the `postinstall` script I guess? :thinking: Not sure we need that script, since it won't run anyway.

File: package.json [hunk 3]
```diff
@@ -52,7 +52,10 @@
     "storybook:build": "build-storybook -c .storybook -o storybook-build --static-dir ./app ./storybook/images",
     "storybook:deploy": "storybook-to-ghpages --existing-output-dir storybook-build --remote storybook --branch master",
     "update-changelog": "./development/auto-changelog.sh",
-    "generate:migration": "./development/generate-migration.sh"
+    "generate:migration": "./development/generate-migration.sh",
+    "lavamoat:auto": "lavamoat ./development/build/index.js --writeAutoPolicy",
+    "lavamoat:debug": "lavamoat ./development/build/index.js --writeAutoPolicyDebug",
+    "postinstall": "yarn patch-package"
```
Comment (kumavis) [inline package.json:58]:
```suggestion
    "lavamoat:debug": "lavamoat ./development/build/index.js --writeAutoPolicyDebug"
```

File: package.json [hunk 4]
```diff
@@ -326,7 +331,8 @@
       "github:assemblyscript/assemblyscript": false,
       "tiny-secp256k1": false,
       "@lavamoat/preinstall-always-fail": false,
-      "fsevents": false
+      "fsevents": false,
+      "metamask-crx": false
```
Comment (kumavis) [inline package.json:335]:
```suggestion
      "fsevents": false
```

**Comment by kumavis** [comment]
This is ready for team review 🎉 🎉 🎉

File: .eslintrc.js [hunk 5]
```diff
@@ -31,6 +31,7 @@ module.exports = {
     'test/e2e/send-eth-with-private-key-test/**',
     'nyc_output/**',
     '.vscode/**',
+    'lavamoat/**',
```
[the remark this answers is not in the document]
Comment (kumavis) [reply .eslintrc.js:34]:
```suggestion
    'lavamoat/*/policy.json',
```

File: .prettierignore [hunk 6]
```diff
@@ -1,4 +1,5 @@
 node_modules/**
+lavamoat/**
```
Comment (kumavis) [inline .prettierignore:2]:
```suggestion
lavamoat/*/policy.json
```

File: patches/graceful-fs+4.2.3.patch [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
++++ b/node_modules/graceful-fs/graceful-fs.js
+@@ -41,42 +41,6 @@ if (!global[gracefulQueue]) {
+     }
+   })
+ 
+-  // Patch fs.close/closeSync to shared queue version, because we need
+-  // to retry() whenever a close happens *anywhere* in the program.
+-  // This is essential when multiple graceful-fs instances are
+-  // in play at the same time.
+-  fs.close = (function (fs$close) {
```
Comment (Gudahtt) [inline patches/graceful-fs+4.2.3.patch:13]: This seems safe as long as we're only using one instance of `graceful-fs` in our build system. Which is true at the moment. I guess we'll just have to keep an eye on this going forward.

File: patches/readable-stream+2.3.6.patch [hunk 8]
```diff
@@ -0,0 +1,103 @@
+diff --git a/node_modules/readable-stream/lib/_stream_duplex.js b/node_modules/readable-stream/lib/_stream_duplex.js
```
Comment (Gudahtt) [inline patches/readable-stream+2.3.6.patch:1]: We don't use v2.3.6 of readable-stream anymore. We use v2.3.7 which includes this patch: https://github.com/nodejs/readable-stream/pull/423
So maybe we don't need this patch at all anymore?

File: patches/stylelint+13.6.1.patch [hunk 9]
```diff
@@ -0,0 +1,28 @@
+diff --git a/node_modules/stylelint/lib/syntaxes/index.js b/node_modules/stylelint/lib/syntaxes/index.js
+index 7afa0c3..73eaa00 100644
+--- a/node_modules/stylelint/lib/syntaxes/index.js
++++ b/node_modules/stylelint/lib/syntaxes/index.js
+@@ -1,16 +1,13 @@
+ 'use strict';
+ 
+-// Use this require pattern so that syntaxes can be bundled separately
+-const importLazy = require('import-lazy')(require);
```
Comment (Gudahtt) [inline patches/stylelint+13.6.1.patch:9]: Interesting :thinking: `stylelint` seems to work OK without this patch. Why was this removed - did this have some impact I'm missing, or was this precautionary?

File: patches/plugin-error+1.0.1.patch [hunk 10]
```diff
@@ -0,0 +1,12 @@
+diff --git a/node_modules/plugin-error/index.js b/node_modules/plugin-error/index.js
+index a4d360d..d2be4a2 100644
+--- a/node_modules/plugin-error/index.js
++++ b/node_modules/plugin-error/index.js
+@@ -54,7 +54,6 @@ function PluginError(plugin, message, options) {
+       return this._messageWithDetails() + '\nStack:';
+     }.bind(this);
+ 
+-    Error.captureStackTrace(safety, arguments.callee || this.constructor);
```
Comment (Gudahtt) [inline patches/plugin-error+1.0.1.patch:9]: The build seems to work fine without this patch as well

File: patches/gulp-livereload++chalk+2.4.2.patch [hunk 11]
```diff
@@ -0,0 +1,47 @@
+diff --git a/node_modules/gulp-livereload/node_modules/chalk/index.js b/node_modules/gulp-livereload/node_modules/chalk/index.js
+index 1cc5fa8..08a1f17 100644
+--- a/node_modules/gulp-livereload/node_modules/chalk/index.js
++++ b/node_modules/gulp-livereload/node_modules/chalk/index.js
+@@ -47,14 +47,7 @@ function Chalk(options) {
+ 	applyOptions(this, options);
+ }
+ 
+-// Use bright blue on Windows as the normal blue color is illegible
```
Comment (Gudahtt) [inline patches/gulp-livereload++chalk+2.4.2.patch:9]: The build seems to work without this patch. This package is used for livereload, so I tested this by running `yarn start` and changing files, to see whether anything broke.

**Review by Gudahtt — COMMENTED** [review]
I tried testing this by deleting the readable-stream patch, updating the graceful-fs patch, and updating the policy file by running `yarn lavamoat --autopolicy ./development/build/index.js prod`. There are lots of new console warnings now during the build :thinking: 

Lots of new Buffer constructor deprecation warnings:
```
[DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
```

and many from LavaMoat about attempts to write read-access globals.

The Buffer warnings appeared to be from packages that we were using before, so I'm unsure why they didn't show up before :thinking:.

I'm guessing the approach taken with the LavaMoat warnings is to assume the write attempts weren't all that important? I'm unsure how to investigate further, as the warning doesn't indicate where the write attempt occurred.

File: patches/stylelint+13.6.1.patch (hunk 9 above)
Comment (kumavis) [reply patches/stylelint+13.6.1.patch:9]: `importLazy` breaks our static analysis. removing this patch will work until we regenerate the policy file

File: patches/plugin-error+1.0.1.patch (hunk 10 above)
Comment (kumavis) [reply patches/plugin-error+1.0.1.patch:9]: this is only a problem when the build system throws an error

File: patches/gulp-livereload++chalk+2.4.2.patch (hunk 11 above)
Comment (kumavis) [reply patches/gulp-livereload++chalk+2.4.2.patch:9]: this one can be removed. latest LavaMoat no longer protects module exports : (

File: patches/graceful-fs+4.2.3.patch (hunk 7 above)
Comment (kumavis) [reply patches/graceful-fs+4.2.3.patch:13]: previously you could not modify the exports of other packages (here its `fs`), but Lavamoat no longer protects module exports. this patch can be removed.

File: patches/plugin-error+1.0.1.patch (hunk 10 above)
Comment (kumavis) [reply patches/plugin-error+1.0.1.patch:9]: this should remain as it is

File: patches/gulp-livereload++chalk+2.4.2.patch (hunk 11 above)
Comment (kumavis) [reply patches/gulp-livereload++chalk+2.4.2.patch:9]: this patch should be removed

File: patches/stylelint+13.6.1.patch (hunk 9 above)
Comment (kumavis) [reply patches/stylelint+13.6.1.patch:9]: this should remain as it is

File: patches/readable-stream+2.3.6.patch (hunk 8 above)
Comment (kumavis) [reply patches/readable-stream+2.3.6.patch:1]: this patch should be updated or removed

File: patches/graceful-fs+4.2.3.patch (hunk 7 above)
Comment (kumavis) [reply patches/graceful-fs+4.2.3.patch:13]: this patch should be removed

File: patches/graceful-fs+4.2.3.patch [hunk 12]
```diff
@@ -0,0 +1,47 @@
+diff --git a/node_modules/graceful-fs/graceful-fs.js b/node_modules/graceful-fs/graceful-fs.js
```
[the remark this answers is not in the document]
Comment (kumavis) [reply patches/graceful-fs+4.2.3.patch:1]: this patch should be updated or removed

**Comment by kumavis** [comment]
should review console warnings

**Comment by Gudahtt** [comment]
I looked into the that "ignoring write attempt" warning, and changed it to throw an error locally to help me debug it. It turns out that the first warning I hit in any process is caused by _dart_. I encounter that in every process, even those that don't touch dart at all (or so I thought).

e.g.

```
$ yarn lavamoat ./development/build/index.js build scripts:core:dev:disable-console
yarn run v1.22.10
$ /home/gudahtt/projects/metamask-extension/node_modules/.bin/lavamoat ./development/build/index.js build scripts:core:dev:disable-console
(Error#1)

  Error#1: LavaMoat: ignoring write attempt to read-access global "process"
     at Object.set (LavaMoat/core/kernel:258:17)
    at Object.eval (/home/gudahtt/projects/metamask-extension/node_modules/sass/sass.dart.js:26:16)
    at internalRequire (LavaMoat/core/kernel:353:25)
    at requireRelative (LavaMoat/core/kernel:394:27)
    at requireRelativeWithContext (LavaMoat/core/kernel:366:16)
    at Object.eval (/home/gudahtt/projects/metamask-extension/development/build/sass-compiler.js:2:14)
    at internalRequire (LavaMoat/core/kernel:353:25)
    at requireRelative (LavaMoat/core/kernel:394:27)
    at requireRelativeWithContext (LavaMoat/core/kernel:366:16)
    at Object.eval (/home/gudahtt/projects/metamask-extension/development/build/styles.js:16:17)

```

**Comment by Gudahtt** [comment]
I was able to remove warnings from all build steps apart from `styles` by making `sass` and `sass-compiler` a dynamic require step. Good to know that all of those warnings were just dart.

**Comment by Gudahtt** [comment]
I'm seeing a bunch of changes after running `yarn lavamoat:auto` :thinking: 
I pushed them here so you could see them: https://github.com/MetaMask/metamask-extension/tree/lavamoat-buildsystem-auto-policy-changes

**Comment by kumavis** [comment]
@Gudahtt the auto policy charges are likely because you made the require dynamic? we dont know to walk it then

**Comment by Gudahtt** [comment]
Sorry I meant from a clean working tree, after discarding the dynamic import change.

I figured it was due to some platform difference (mac vs Linux)

**Comment by kumavis** [comment]
@Gudahtt ah yeah may be optional deps diff due to platform. @EtDu is running mac and gets fsevents

**Review by Gudahtt — APPROVED** [review]
I was able to confirm that the policy differences were due to platform; the changes were all related to `fsevents`.

These cross-platform issues do present challenges to keeping the auto policy up-to-date. We'll have to postpone adding a check in CI to ensure it has been updated, as it'll always fail for contributors using macOS (and possibly Windows). But we can make do for the time being by checking prior to each release to ensure the policy file is up-to-date. I can create a CI check in a separate PR to run on release branches only, to ensure we don't accidentally forget this step.

Similarly, we can address the various console warnings in a separate PR.

LGTM!
